java/tools/src/main/java/org/apache/arrow/tools/Integration.java

```diff
@@ -247,14 +248,23 @@ private static void compare(VectorSchemaRoot arrowRoot, VectorSchemaRoot jsonRoo
       for (int j = 0; j < valueCount; j++) {
         Object arrow = arrowVector.getAccessor().getObject(j);
         Object json = jsonVector.getAccessor().getObject(j);
-        if (!Objects.equal(arrow, json)) {
+        if (!equals(field.getType(), arrow, json)) {
           throw new IllegalArgumentException(
               "Different values in column:\n" + field + " at index " + j + ": " + arrow + " != " + json);
         }
       }
     }
   }
 
+  private static boolean equals(ArrowType type, final Object arrow, final Object json) {
```

Comment:
See my proposed functions in the jira. The code here is insufficient for a number of situations.